Documentation/update-modules-v3-file-api.md

```diff
@@ -0,0 +1,422 @@
+followed the update module may hang.
+
+**Important:** An update module must not install the update in the final
+location during the streaming stage, because checksums are not verified until
+after the streaming stage is over. If it must be streamed to the final location
+(such as for example a partition), it should be stored in an inactive state, so
+that it is not accidentally used, and then be activated in the
+`ArtifactInstall` stage. Failure to do so can mean that the update module will
+be vulnerable to security attacks.
+
+### File tree
+
+The file tree only exists in the `ArtifactInstall` and later states, and only
+if the streams were **not** consumed during the `Download` state. In this case
+Mender will download the streams automatically and put them in the file tree.
+
+The `files` directory contains the payloads from the Artifact, and is taken from
+the `data/nnnn.tar.gz` payload that corresponds to the indexed subfolder being
+processed by Mender, just like the header.
+
+
+Execution
+---------
+
+Since the API may be expanded in the future with additional calls and states,
+all update modules should simply print nothing and return zero if it is invoked
+with an unknown first argument.
+
+For all the states, the update module is called once for each state that occurs,
+with the working directory set to the directory where the File API resides. It
+is called with exactly two arguments: The current state, and the absolute path
+of the File API location.
+
+Returning any non-zero value in the update module triggers a failure, and will
+invoke the relevant failure states.
+
+
+Signatures and augmented artifacts
+----------------------------------
+
+**Warning:** Augmented artifacts are by their very nature security sensitive,
+and it is easy to open up for vulnerabilities if the consequences are not fully
+understood. It is recommended not to use augmented artifacts unless strictly
+needed, and not until the reader has a solid understanding of how it works.
+
+If signatures are being used, sometimes it may be necessary to put data into the
+artifact that isn't signed, while at the same time keeping a trusted chain. For
+example, one can generate a master image, which is signed, and then generate
+many binary delta images from this. Ideally one would not like to sign each and
+every one of these.
+
+Augmented artifacts are artifacts that can contain some parts that are signed
+(the "original" part) and some parts that are not (the "augmented"
+part). Obviously the unsigned content can be very security sensitive, and by
+default, all content in these files will be rejected. Only consider enabling
+augmented content if your update module is prepared to handle untrusted input
+and still guarantee a trusted result (more about best practices below).
+
+### Marking augmented artifacts as supported
+
+If you wish to accept `augment` files you need to implement this calling
+interface:
+
+```bash
+./update-module SupportsAugmentedArtifacts
```

Comment:
I wonder if this is potentially dangerous or could lead to unexpected behavior for update modules that don't support augmented..?
If update-module only supports "install", for example and assumes it would only be called with that, could it install / make some system changes that were not intended?

Maybe a bit far-fetched example as it should be able to support other stages, but just a bit concerned what could happen if the module is not implemented "safely" and then it gets called (as root) with an argument it doesn't support..

I don't really have a good proposal to detect this in run-time, it feels more like it should be part of some "manifest" or "registry" of update modules. Right now it could just state the update module and whether it supports augments or not, in the future it can be extended to check for rollback capability, reboot behavior, etc.

Reply:
Handling unknown arguments is [a required part of the spec](https://github.com/mendersoftware/mender/pull/330/files#diff-1122e6d8a5474b1546480d6943246a7aR192) and is important for being able to extend update modules in the future. Besides, all update modules will be called with a minimum of four different arguments: Download, ArtifactInstall, ArtifactCommit and Cleanup. So they need to be prepared for that, even if they only implement one of them.

Reply:
> I don't really have a good proposal to detect this in run-time, it feels more like it should be part of some "manifest" or "registry" of update modules. Right now it could just state the update module and whether it supports augments or not, in the future it can be extended to check for rollback capability, reboot behavior, etc.

One reason for not using a manifest: When doing package installs, you often don't know whether you need a reboot until after the package is installed. For example, kernel updates will set a flag, which you can see in your UI as "your machine needs to reboot". So I think this should be dynamic, with Mender asking after the install stage: "Do you need to reboot now?"

Reply:
OK! If you think it's safe then let's keep it that way. The simpler, the better. :)